### DISCUSSION TITLE
Optional reducer migrations

### DISCUSSION THREAD

**Discussion by natemann** [pointfreeco/swift-composable-architecture#1291] [Beta]
A majority of my reducers are optional. Looking at the docs and migration guides, I do not see how to convert an optional ReducerProtocol into an AnyRuder for migration purposes. Has anyone been successful in doing this?

**Comment by jshier** [comment]
Yes, you an adapt your new reducer into an `AnyReducer` and use the same `optional` and `pullback` modifiers you did before. 

In my project, something like this:

```swift
expenseDomainReducer
    .pullback(state: \Identified.value, action: .self)
    .optional()
    .pullback(state: \SignedInState.selectedExpense,
              action: /SignedInAction.expense,
              environment: { ExpenseDomain.Environment(api: $0.api) })
```

became this:

```swift
AnyReducer { environment in
    ExpenseDomain(environment: .init(api: environment.api))
}
.pullback(state: \Identified.value, action: .self)
.optional()
.pullback(state: \SignedInState.selectedExpense,
          action: /SignedInAction.expense,
          environment: { ExpenseDomain.Environment(api: $0.api) })
```

**Reply by stephencelis** [reply]
In addition, when you fully migrate away from `AnyReducer` (it's soft-deprecated), you will end up with something like:

```swift
struct SignedIn: ReducerProtocol {
  struct State {
    // ...
  }
  enum Action {
    // ...
  }
  // ...
  var body: some ReducerProtocol<State, Action> {
    Reduce { state, action in
      // core feature logic
    }
    .ifLet(\.selectedExpense, action: /Action.expense) {
      Scope(state: \Identified.value, action: .self) {
        ExpenseDomain() // API propagated via @Dependency
      }
    }
  }
}
```

Navigation in this example is definitely clunky at the moment, but we have some very promising tools in development!

**Comment by mbrandonw** [comment]
We just updated our [migration article][migration-docs] to describe how to step-by-step migrate optional+pullback reducers:

Please let us know if you have any feedback on that.

[migration-docs]: https://github.com/pointfreeco/swift-composable-architecture/blob/protocol/Sources/ComposableArchitecture/Documentation.docc/Articles/MigratingToReducerProtocols.md#optional-and-pullback-reducers

**Reply by jshier** [reply]
May want to include environment passing or extracting values from the environment in that example, given your other examples move them into properties. Or just ignore the `environment` value in the example.

**Reply by mbrandonw** [reply]
Ah yes, good point. Just updated the doc, thanks!

**Comment by natemann** [comment]
These examples only seem to work if the parent and child ReducerProtocols are in the same module. When the child ReducerProtocol is in a separate module, things do not compile. For example 
```swift 
public static let reducer = AnyReducer<State, Action, Environment>.combine(

AnyReducer {
    RevenueSplitList(databaseClient: $0.databaseClient, uuid: $0.uuid)
}.optional()
    .pullback(
      state: \.revenueSplitList,
      action: /Action.revenueSplitList,
      environment: { 
        RevenueSplitList(
          databaseClient: $0.databaseClient,
          uuid: $0.uuid) }),
```
Does not compile. The compile thinks $0 in `AnyReducer` is `RevenueSplitList`. Even casting as `Environment` fails.

**Reply by mbrandonw** [reply]
According to this line it seems like you are turning the environment into `RevenueSplitList`:

```swift
      environment: { 
        RevenueSplitList(
```

Is that supposed to be `RevenueEnvironment` or something like that?

**Reply by natemann** [reply]
well, that is another question.  RevenueSplitList is a reducerProtocol, so it does not have an explicit Environment anymore.

**Reply by natemann** [reply]
I know going to Dependencies will probably solve all this, but we are not at that point yet

**Reply by jshier** [reply]
@natemann `AnyReducer` provides the environment in the initializer closure so there's no need to do anything to get it.

**Reply by mbrandonw** [reply]
Oh good point, that's pointing out a problem in our migration guide. By the time you have gotten to this point in the refactor there is no more child environment, and so you can actually just use a tuple:

```swift
environment: { 
  (
    databaseClient: $0.databaseClient,
    uuid: $0.uuid
  )
},
```

Then the `RevenueSplitList` can pluck off what it needs from the tuple.

We actually do quite a bit of reducer migration work in next week's episode, so hopefully seeing us do it live will help you see how these things go in complex applications. I will also update the migrate strategy to mention this.

**Reply by mbrandonw** [reply]
Oh yeah, and actually what @jshier is saying is even simpler. Just use `environment: { $0 }` to pass along the parent environment, and then inside `AnyReducer` you can pluck out whatever you need.

**Reply by natemann** [reply]
ok, this works
```swift 
AnyReducer {
      RevenueSplitList(databaseClient: $0.1, uuid: $0.2)
    }.optional()
      .pullback(
        state: \.revenueSplitList,
        action: /Action.revenueSplitList,
        environment: { ($0.databaseClient, $0.uuid) }),
```
So is it correct to say the `environment` variable in the `AnyReducer` comes from the pullback environment

**Reply by natemann** [reply]
@mbrandonw yes, I think I just figured that out
